### ISSUE TITLE
SignatureDoesNotMatch exception when using custom service endpoint that has a path component.

### ISSUE THREAD

**Issue by rafd123** [aws/aws-sdk-net#240]
Here's a quick snippet that demonstrates how one can reproduce the issue; note that the service URL has a path component:

    var clientConfig = new AmazonEC2Config
    {
        ServiceURL = "http://somehost/somepath"
    };

    var ec2Client = new AmazonEC2Client(
            new BasicAWSCredentials("access_key_here", "access_key_secret_here"),
            clientConfig);

    var response = ec2Client.DescribeAccountAttributes();

The root cause appears to be that when AWS4Signer goes to canonicalize the request, it strictly uses IRequest.ResourcePath, which is null most of the time; in some cases, it gets hardcoded to "/". As a result, the canonicalized request has the wrong canonical URL which ultimately leads to the request being signed incorrectly.

**Comment by rafd123** [comment]
@PavelSafronov I'm a little confused why this is labeled as a feature request when the issue seems to be a violation of the [Signature Version 4 spec](http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html):

> The canonical URI is the URI-encoded version of the absolute path component of the URI—everything in the URI from the HTTP host to the question mark character ("?") that begins the query string parameters (if any).

> Normalize URI paths according to RFC 3986 by removing redundant and relative path components. Each path segment must be URI-encoded. If the absolute path is empty, use a forward slash (/). In the example request, nothing follows the host in the URI, so the absolute path is empty.

Additionally, other AWS SDKs (e.g. boto3) handle this case with no issues.

**Comment by PavelSafronov** [comment]
Currently, the .NET SDK does not access any AWS services which use subpaths in their endpoints, so I judged that the lack of this feature was not strictly a bug but rather a feature. That being said, I have no objection to labeling this as a bug if the lack of this functionality is causing issues for you.

**Comment by steveataws** [comment]
PR merged and updated core version 3.1.1.1 released, closing.

**Comment by PavelSafronov** [comment]
@rafd123 it looks like this commit has broken a [fairly specific scenario](https://github.com/aws/aws-sdk-net/issues/259), so I'm looking at reworking the approach we use to access a service with a path component. Can you provide some information on how you are accessing a service with a path component? I'm trying to figure out how to test your specific scenario so I can verify that my alternative approach doesn't end up breaking you.

**Comment by PavelSafronov** [comment]
The .NET SDK team is planning on making the following change to AWS4Signer, which may end up breaking this specific scenario. **Please notify us as soon as possible if it does.** Our current plan is to make and release this fix by November 17th, 2015.

```
protected static string CanonicalizeResourcePath(Uri endpoint, string resourcePath)
{
    if (string.IsNullOrEmpty(resourcePath))
        return Slash;

    if (endpoint != null)
    {
        var path = endpoint.AbsolutePath;
        if (string.IsNullOrEmpty(path) || string.Equals(path, Slash, StringComparison.Ordinal))
            path = string.Empty;

        if (resourcePath.StartsWith(Slash, StringComparison.Ordinal))
            resourcePath = resourcePath.Substring(1);

        resourcePath = path + Slash + resourcePath;
    }

    // split path at / into segments
    var pathSegments = resourcePath.Split(new char[] { SlashChar }, StringSplitOptions.None);

    // url encode the segments
    var encodedSegments = pathSegments
        .Select(segment => AWSSDKUtils.UrlEncode(segment, false))
        .ToArray();

    // join the encoded segments with /
    var canonicalizedResourcePath = string.Join(Slash, encodedSegments);

    return canonicalizedResourcePath;
}
```

**Comment by rafd123** [comment]
The proposed code will break the scenario of a custom endpoint that has a path component. In it, if `resourcePath` is null or empty (which is true for EC2 calls, for example), the endpoint's path component never has an opportunity to get included as part of the canonicalized path.

I think modifying the proposed code as follows solves both issues for the version 4 signer:

        protected static string CanonicalizeResourcePath(Uri endpoint, string resourcePath)
        {
            if (endpoint != null)
            {
                var path = endpoint.AbsolutePath;
                if (string.IsNullOrEmpty(path) || string.Equals(path, Slash, StringComparison.Ordinal))
                    path = string.Empty;

                if (!string.IsNullOrEmpty(resourcePath) && resourcePath.StartsWith(Slash, StringComparison.Ordinal))
                    resourcePath = resourcePath.Substring(1);

                if (!string.IsNullOrEmpty(resourcePath))
                    path = path + Slash + resourcePath;

                resourcePath = path;
            }

            if (string.IsNullOrEmpty(resourcePath))
                return Slash;

            // split path at / into segments
            var pathSegments = resourcePath.Split(new char[] { SlashChar }, StringSplitOptions.None);

            // url encode the segments
            var encodedSegments = pathSegments
                .Select(segment => AWSSDKUtils.UrlEncode(segment, false))
                .ToArray();

            // join the encoded segments with /
            var canonicalizedResourcePath = string.Join(Slash, encodedSegments);

            return canonicalizedResourcePath;
        }

FWIW, I looked into why 608fdcf76078abc11893dd7b4cdb6d1c10c3fbb4 regressed WP 8.1. The original intent was to sign the request with the same URL that was to be used to make the request by calling `AmazonServiceClient.ComposeUrl()` to ensure parity between the contents of the signature and the URL the server will ultimately use to recreate the signature. `AmazonServiceClient.ComposeUrl()` ends up calling `AmazonServiceClient.DontUnescapePathDotsAndSlashes()` in an attempt to smooth over differences in the way the Uri class behaves across .NET versions; clearly this is a no-op on WP 8.1 and of course for WP 8.1 the Uri class behaves in the "old" way by unescaping the slashes. What I don't understand is how the final URL for the request is composed before it heads out the door on WP 8.1, and how is it getting away without unescaping the slashes.

**Comment by rafd123** [comment]
@PavelSafronov  I submitted a pull request with the changes you suggested along with unit tests to document the intend.

The reason the "custom endpoint with path component" scenario is important to us is because our group is trying to wean existing users off of other hypervisor technologies and use AWS as a standard. We have an AWS adapter in front of our stack to facilitate this, but due to technical restrictions, the endpoint hosted by this adapter needs a path component. The other SDKs we've tested worked out of the box (e.g. boto, the ruby SDK); we just wanted to ensure the .NET SDK didn't pose as an impedance.

To answer you question: the ultimate way this was being tested was to point an AWS client at our custom endpoint (see example in the originating comment of this issue) and ensuring that the server didn't return a SignatureDoesNotMatch error.

Ultimately, it came down to ensuring the endpoint's path components are represented in the canonicalized URL per the signing spec.
